File: pkg/integrations/github/webhook_handler.go
package github

import (
	"context"
	"fmt"

	"github.com/google/go-github/v74/github"
	"github.com/mitchellh/mapstructure"
	"github.com/superplanehq/superplane/pkg/core"
)

type WebhookConfiguration struct {
	EventType  string   `json:"eventType"`
	EventTypes []string `json:"eventTypes"` // Multiple event types (takes precedence over EventType if set)
	Repository string   `json:"repository"`
}

type Webhook struct {
	ID          int64  `json:"id"`
	WebhookName string `json:"name"`
}

type GitHubWebhookHandler struct{}

func (h *GitHubWebhookHandler) CompareConfig(a, b any) (bool, error) {
	configA := WebhookConfiguration{}
	configB := WebhookConfiguration{}

	err := mapstructure.Decode(a, &configA)
	if err != nil {
		return false, err
	}

	err = mapstructure.Decode(b, &configB)
	if err != nil {
		return false, err
	}

	if configA.Repository != configB.Repository {
		return false, nil
	}

	// Compare event types - normalize to slices for comparison
	eventsA := configA.EventTypes
	if len(eventsA) == 0 && configA.EventType != "" {
		eventsA = []string{configA.EventType}
	}

	eventsB := configB.EventTypes
	if len(eventsB) == 0 && configB.EventType != "" {
		eventsB = []string{configB.EventType}
	}

	if len(eventsA) != len(eventsB) {
		return false, nil
	}

	// Create a map to compare events regardless of order
	eventMap := make(map[string]bool)
	for _, e := range eventsA {
		eventMap[e] = true
	}
	for _, e := range eventsB {
		if !eventMap[e] {
			return false, nil
		}
	}

	return true, nil
}

func (h *GitHubWebhookHandler) Merge(current, requested any) (any, bool, error) {
	return current, false, nil
}

func (h *GitHubWebhookHandler) Setup(ctx core.WebhookHandlerContext) (any, error) {
	metadata := Metadata{}
	err := mapstructure.Decode(ctx.Integration.GetMetadata(), &metadata)
	if err != nil {
		return nil, err
	}

	client, err := NewClient(ctx.Integration, metadata.GitHubApp.ID, metadata.InstallationID)
	if err != nil {
		return nil, err
	}

	config := WebhookConfiguration{}
	err = mapstructure.Decode(ctx.Webhook.GetConfiguration(), &config)
	if err != nil {
		return nil, err
	}

	secret, err := ctx.Webhook.GetSecret()
	if err != nil {
		return nil, fmt.Errorf("error getting webhook secret: %v", err)
	}

	// Use EventTypes if set, otherwise fall back to single EventType
	events := config.EventTypes
	if len(events) == 0 && config.EventType != "" {
		events = []string{config.EventType}
	}

	hook := &github.Hook{
		Active: github.Ptr(true),
		Events: events,
		Config: &github.HookConfig{
			URL:         github.Ptr(ctx.Webhook.GetURL()),
			Secret:      github.Ptr(string(secret)),
			ContentType: github.Ptr("json"),
		},
	}

	createdHook, _, err := client.Repositories.CreateHook(context.Background(), metadata.Owner, config.Repository, hook)
	if err != nil {
		return nil, fmt.Errorf("error creating webhook: %v", err)
	}

	return &Webhook{ID: createdHook.GetID(), WebhookName: *createdHook.Name}, nil
}

func (h *GitHubWebhookHandler) Cleanup(ctx core.WebhookHandlerContext) error {
	metadata := Metadata{}
	err := mapstructure.Decode(ctx.Integration.GetMetadata(), &metadata)
	if err != nil {
		return err
	}

	client, err := NewClient(ctx.Integration, metadata.GitHubApp.ID, metadata.InstallationID)
	if err != nil {
		return err
	}

	webhook := Webhook{}
	err = mapstructure.Decode(ctx.Webhook.GetMetadata(), &webhook)
	if err != nil {
		return err
	}

	configuration := WebhookConfiguration{}
	err = mapstructure.Decode(ctx.Webhook.GetConfiguration(), &configuration)
	if err != nil {
		return err
	}

	_, err = client.Repositories.DeleteHook(context.Background(), metadata.Owner, configuration.Repository, webhook.ID)
	if err != nil {
		return fmt.Errorf("error deleting webhook: %v", err)
	}

	return nil
}

File: extensions/examples/github/src/integrations/webhook-handler.ts
import type { IntegrationWebhookHandler, RuntimeValue } from "@superplanehq/sdk";
import { githubRequest, type GitHubWebhook, parseRepositoryOwnerAndName, readGitHubToken } from "../lib/github.js";

interface WebhookConfiguration {
  eventType?: string;
  eventTypes?: string[];
  repository?: string;
}

interface ProvisionedWebhookMetadata {
  id: number;
  webhookName: string;
}

export const githubWebhookHandler = {
  compareConfig({ current, requested }) {
    const normalizedCurrent = normalizeWebhookConfiguration(current);
    const normalizedRequested = normalizeWebhookConfiguration(requested);

    if (normalizedCurrent.repository !== normalizedRequested.repository) {
      return false;
    }

    const currentEvents = normalizeEventTypes(normalizedCurrent);
    const requestedEvents = normalizeEventTypes(normalizedRequested);
    if (currentEvents.length !== requestedEvents.length) {
      return false;
    }

    const currentSet = new Set(currentEvents);
    return requestedEvents.every((event) => currentSet.has(event));
  },
  merge({ current }) {
    return {
      merged: current,
      changed: false,
    };
  },
  async setup({ runtime, webhook }) {
    const config = normalizeWebhookConfiguration(await webhook.getConfiguration());
    const token = await readGitHubToken(runtime.integration);
    const { owner, repository } = parseRepositoryOwnerAndName(config.repository ?? null);
    const secret = await webhook.getSecret();
    const events = normalizeEventTypes(config);

    const hook = await githubRequest<GitHubWebhook>(runtime.http, {
      method: "POST",
      path: `/repos/${owner}/${repository}/hooks`,
      token,
      body: {
        active: true,
        events,
        config: {
          url: await webhook.getURL(),
          secret: new TextDecoder().decode(secret),
          content_type: "json",
        },
      },
    });

    return {
      id: hook.id,
      webhookName: hook.name,
    };
  },
  async cleanup({ runtime, webhook }) {
    const config = normalizeWebhookConfiguration(await webhook.getConfiguration());
    const metadata = normalizeProvisionedWebhookMetadata(await webhook.getMetadata());
    if (!metadata) {
      return;
    }

    const token = await readGitHubToken(runtime.integration);
    const { owner, repository } = parseRepositoryOwnerAndName(config.repository ?? null);
    await githubRequest<unknown>(runtime.http, {
      method: "DELETE",
      path: `/repos/${owner}/${repository}/hooks/${metadata.id}`,
      token,
    });
  },
} satisfies IntegrationWebhookHandler;

function normalizeWebhookConfiguration(value: RuntimeValue | null | undefined): WebhookConfiguration {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }

  return value as WebhookConfiguration;
}

function normalizeEventTypes(configuration: WebhookConfiguration): string[] {
  if (Array.isArray(configuration.eventTypes) && configuration.eventTypes.length > 0) {
    return configuration.eventTypes;
  }

  if (typeof configuration.eventType === "string" && configuration.eventType) {
    return [configuration.eventType];
  }

  return [];
}

function normalizeProvisionedWebhookMetadata(value: RuntimeValue): ProvisionedWebhookMetadata | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return null;
  }

  const id = "id" in value ? value.id : undefined;
  const webhookName = "webhookName" in value ? value.webhookName : undefined;
  if (typeof id !== "number" || !Number.isInteger(id)) {
    return null;
  }

  return {
    id,
    webhookName: typeof webhookName === "string" ? webhookName : "",
  };
}
